LoRaEngine/LoraKeysManagerFacade/EdgeDeviceGetter.cs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoraKeysManagerFacade
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using LoRaTools;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Logging;
+
+    internal class EdgeDeviceGetter
+    {
+        private readonly IDeviceRegistryManager registryManager;
+        private readonly ILoRaDeviceCacheStore cacheStore;
+        private readonly ILogger<EdgeDeviceGetter> logger;
+
+        public EdgeDeviceGetter(IDeviceRegistryManager registryManager,
+                                ILoRaDeviceCacheStore cacheStore,
+                                ILogger<EdgeDeviceGetter> logger)
+        {
+            this.registryManager = registryManager;
+            this.cacheStore = cacheStore;
+            this.logger = logger;
+        }
+
+        private async Task<List<Twin>> GetEdgeDevicesAsync()
+        {
+            this.logger.LogDebug("Getting Azure IoT Edge devices");
+            var q = this.registryManager.CreateQuery("SELECT * FROM devices where capabilities.iotEdge = true");
+            var twins = new List<Twin>();
+            do
+            {
+                twins.AddRange(await q.GetNextAsTwinAsync());
+            } while (q.HasMoreResults);
+            return twins;
+        }
+
+        internal async Task<bool> IsEdgeDeviceAsync(string lnsId)
+        {
+            var keyLock = $"{nameof(EdgeDeviceGetter)}-lock";
+            var owner = nameof(EdgeDeviceGetter);
+            var isEdgeDevice = false;
+            try
+            {
+                if (await cacheStore.LockTakeAsync(keyLock, owner, TimeSpan.FromSeconds(10)))
+                {
+                    var findInCache = () => this.cacheStore.GetObject<DeviceKind>(lnsId);
+                    if (findInCache() is null)
+                    {
+                        await RefreshEdgeDevicesCacheAsync();
```

Comment:
One risk here is that if we have a lot of instances of the lns running, it could be that all of them would cause a cache refresh. We may want to introduce a timestamp of the last refresh to mitigate that risk.

Reply:
Done with 281263591b5844d0b0cdb98a04948560e320ddfd
Please check if ok 👍